### DISCUSSION TITLE
Substring Filtering in Altair

### DISCUSSION THREAD

**Discussion by foufoulides** [vega/altair#2547] [Q&A]
Hi all, 

I need help with substring filtering in `transform_filter()` in an Altair chart. 

* I have a dataset with data on various locations.  
* Each location has its own comparable locations group (**so it is *not* a partition of the locations where each location belongs in one and only one group**).  
* I want the user to be able to filter by each location's comparable group. 
* I looked at Vega string expressions and I think using the the `indexof()` function in expression `indexof(string, substring)>-1` (because it returns -1, when the substring is not found in the string) would do it, if I create a list of strings where each string is the concatenation of the locations in each comparable location group, with a dash separator. 
*  For some reason though it does not work and I get a black image, though it works when I hardcode in the function one comparable locations group/string.  
* Can someone help with what I am doing wrong?  A fully coded example is in the image below. 

Thank you in advance, 
Chris

<img width="1000" alt="altair_example" src="https://user-images.githubusercontent.com/88317797/148971086-97deaf6a-d0fe-4eb0-b3a2-3ceb89122546.png">

Code copied from comment below
```python
import pandas as pd
import altair as alt  

# Example DataFrame 
df = pd.DataFrame({'lctn': ['AAA', 'BBB', 'CCC', 'DDD', 'EEE', 'AAA', 'BBB', 'CCC', 'DDD', 'EEE'], 
                   'xval': [10, 4, 1, 9, 7, 1, 3, 4, 2, 8], 
                   'yval': [8, 7, 10, 9, 2, 6, 4, 3, 5, 1]} )

group_dropdown = alt.binding_select(options = sorted(['AAA-BBB', 'BBB-AAA-CCC', 'CCC-BBB-DDD', 'DDD-CCC-EEE', 'EEE-DDD']), labels=list(sorted(df["lctn"].dropna().unique())), name='Group') 
group_selection = alt.selection_single(name='group_sel', fields=['lctn'], bind=group_dropdown)

# This does not work for some reason 
alt.Chart(df).mark_point().encode(
    x=alt.X('xval'),
    y=alt.Y('yval'),
).add_selection(group_selection).transform_filter('indexof(group_sel.lcnt, datum.lctn)>-1').properties(width=80, height=80)

# Hardcoding one comparative groups location string works however 
alt.Chart(df).mark_point().encode(
    x=alt.X('xval'),
    y=alt.Y('yval'),
).add_selection(group_selection).transform_filter('indexof("BBB-AAA-CCC", datum.lctn)>-1').properties(width=80, height=80)
```

**Comment by mjohnson11** [comment]
Try writing `group_sel.lctn[0]` in the expression. This confused me a lot because I was just working on the opposite problem and `indexof(datum.lctn, group_sel.lctn)>-1` works without the [0] index - which I think is just because of how indexof works in javascript: `'aaa-bbb'.indexOf(['bbb'])` will work and yield 4. 

Sorry if it was my question/answer on stack overflow that led you down the wrong path! Will update it!

**Reply by foufoulides** [reply]
Hi Milo, 

Thank you for the prompt reply. I did find your post on stack overflow after not being able to make this work and I noticed that you made it work the other way around. So I could see I was on the right track, but couldn't understand why the opposite wouldn't work!  I just tried what you suggested, but unfortunately the issue is still there. 

Best, 
Chris

**Comment by jakevdp** [comment]
Can you please copy and paste code as text into your question rather than pasting a screenshot of code?

**Reply by foufoulides** [reply]
Hi Jake, 

Thank you for the prompt reply.  

```
import altair as alt  

# Example DataFrame 
df = pd.DataFrame({'lctn': ['AAA', 'BBB', 'CCC', 'DDD', 'EEE', 'AAA', 'BBB', 'CCC', 'DDD', 'EEE'], 
                   'xval': [10, 4, 1, 9, 7, 1, 3, 4, 2, 8], 
                   'yval': [8, 7, 10, 9, 2, 6, 4, 3, 5, 1]} )

group_dropdown = alt.binding_select(options = sorted(['AAA-BBB', 'BBB-AAA-CCC', 'CCC-BBB-DDD', 'DDD-CCC-EEE', 'EEE-DDD']), 
                                    labels=list(sorted(df["lctn"].dropna().unique())), name='Group') 
group_selection = alt.selection_single(name='group_sel', fields=['lctn'], bind=group_dropdown)

# This does not work for some reason 
alt.Chart(df).mark_point().encode(
    x=alt.X('xval'),
    y=alt.Y('yval'),
).add_selection(group_selection).transform_filter('indexof(group_sel.lctn, datum.lctn)>-1').properties(width=80, height=80)

# Hardcoding one comparative groups location string works however 
alt.Chart(df).mark_point().encode(
    x=alt.X('xval'),
    y=alt.Y('yval'),
).add_selection(group_selection).transform_filter('indexof("BBB-AAA-CCC", datum.lctn)>-1').properties(width=80, height=80)
```
Best, 
Chris

**Reply by mjohnson11** [reply]
Just a typo: `lcnt` instead of `lctn` - I think this is a tough part about using expressions right now - no error comes through telling you that. If you fix that and do the `[0]` indexing, it should work! I think the takeaway here is: selectionName.field returns an array in expressions.

**Reply by mjohnson11** [reply]
Just checked and these expression errors don't come through in vega-lite either, so if that improvement is ever made it would be there, not in altair. In general, I think whether and how javascript throws (or doesn't throw) these kinds of errors would probably make figuring out this kind of error handling pretty hard.

**Reply by foufoulides** [reply]
Hi Milo, 

Thanks for spotting the typo!  I actually changed the field name before submitting this here and I typed it wrong at that spot. Also fixed the typo above. I just tried your solution again. It takes a while to work for some reason, but it does work. The graph comes a bit "messed up" on the axes as seen below however.
<img width="500" alt="Capture" src="https://user-images.githubusercontent.com/88317797/148999619-45bf246e-da4e-4ddb-812e-c9394930ad48.PNG">

**Reply by mjohnson11** [reply]
Try including `init={'lctn': 'AAA-BBB'}` in the selection definition. I'm not sure why it behaves strangely when init is not set.

**Reply by foufoulides** [reply]
That worked, thank you so much!  Now that you mentioned this, I noticed in other instances as well that filters often don't work exactly right when initial/default value is not set, but not having such a big impact on the visual. They just need to be reselected to actually filter.

**Reply by mjohnson11** [reply]
No problem! I'm also currently learning how to use expressions in altair - powerful but tough to debug! On the selection initialization, the other important thing to pay attention to is probably the `empty` argument: https://altair-viz.github.io/user_guide/generated/api/altair.selection_single.html
